src/tangents.jl

```diff
@@ -436,16 +440,89 @@ end
     tangent_field_zeros_exprs = ntuple(fieldcount(P)) do n
         if tangent_field_type(P, n) <: PossiblyUninitTangent
             V = PossiblyUninitTangent{tangent_type(fieldtype(P, n))}
-            return :(isdefined(x, $n) ? $V(zero_tangent(getfield(x, $n))) : $V())
+            return :(isdefined(x, $n) ? $V(zero_tangent_internal(getfield(x, $n))) : $V())
         else
-            return :(zero_tangent(getfield(x, $n)))
+            return :(zero_tangent_internal(getfield(x, $n)))
         end
     end
     backing_data_expr = Expr(:call, :tuple, tangent_field_zeros_exprs...)
     backing_expr = :($(backing_type(P))($backing_data_expr))
     return :($(tangent_type(P))($backing_expr))
 end
 
+@inline zero_tangent_internal(x::Union{Int8,Int16,Int32,Int64,Int128,IEEEFloat}, ::IdDict) = zero_tangent_internal(x)
+@inline function zero_tangent_internal(x::SimpleVector, stackdict::IdDict)
+    return map!(n -> zero_tangent_internal(x[n], stackdict), Vector{Any}(undef, length(x)), eachindex(x))
+end
+@inline function zero_tangent_internal(x::Array{P, N}, stackdict::IdDict) where {P, N}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+    if isbitstype(P)
+        return _map_if_assigned!(Base.Fix2(zero_tangent_internal, stackdict), Array{tangent_type(P), N}(undef, size(x)...), x)
+    else
+        stackdict[x] = Array{tangent_type(P), N}(undef, size(x)...)
+        for i in eachindex(x)
+            if isassigned(x, i)
+                stackdict[x][i] = zero_tangent_internal(x[i], stackdict)
+            end
+        end
+        return stackdict[x]
+    end
+end
+function zero_tangent_internal(x::P, stackdict::IdDict) where {P<:Union{Tuple, NamedTuple}}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
+    if tangent_type(P) == NoTangent
+        return NoTangent()
+    end 
+    stackdict[x] = tuple_map(Base.Fix2(zero_tangent_internal, stackdict), x)
+    return stackdict[x]
+end
+function zero_tangent_internal(x::P, stackdict::IdDict) where {P}
+    if haskey(stackdict, x)
+        return stackdict[x]
+    end
```

Comment:
Should this maybe live inside the `tangent_type(P) <: MutableTangent` if statement? It'll be correct either way, but where it currently is, I imagine it'll damage the performance if `tangent_type(P) == NoTangent`, or is a `struct`.